Adds pinecone metadata and sparse values to written vector

alarbada opened the pull request:
### Description

Stacked from #15 

- Adds pinecone metadata
- Adds sparse values

hariso commented on writer.go:
```diff
@@ -114,27 +114,42 @@ func recordID(Key sdk.Data) string {
 }
 
-func recordPayload(payload sdk.Change) ([]float32, error) {
+type recordPayload struct {
```
IIUC, this is more like pineconePayload?

hariso replied:
:point_up:

hariso commented on writer.go:
```diff
@@ -114,27 +114,42 @@ func recordID(Key sdk.Data) string {
+
+func (r recordPayload) PineconeSparseValues() *pinecone.SparseValues {
+	v := &pinecone.SparseValues{r.SparseValues.Indices, r.SparseValues.Values}
+	return v
```
It looks like the method might not be needed, as you're simply returning a value.

alarbada commented on writer.go:
```diff
@@ -114,27 +114,42 @@ func recordID(Key sdk.Data) string {
 	convertedMap := make(map[string]interface{})
 	for key, value := range data {
-		convertedMap[key] = value
+		if trimmed, hasPrefix := trimPineconeKey(key); hasPrefix {
```
According to https://docs.pinecone.io/guides/data/filtering-with-metadata

pinecone metadata is used to allow custom filtering. My reasoning to exclude conduit specific metadata was that the connector user won't care about grouping vectors by which opencdc version were they written, from which plugin where they written, etc. The readAt / createdAt conduit metadata fields might be useful for grouping, idk.

It felt to me a bit dirty to merge conduit metadata / pinecone metadata in a single pinecone vector.

alarbada replied:
@hariso but now that you mention it, it could be useful for the connector user to have a way to distinguish vectors written by conduit.

If I remove the filtering it keeps the connector simple, so I'm up for removing this logic.

hariso replied:
Yeah, we can do that for now. A user can always remove metadata in a processor. Or, if we see many requests by users to remove the metadata, we'll change the connector.:)

lovromazgon replied:
If you ask me, we should stay away from assumptions. Even though this might seem user-friendly and something that the user might want to do anyway, I'd argue that Conduit gets harder to use in general, if I have to check out every connector and its specific user-friendly logic. My expectation is that if a record contains 3 fields of metadata, I want to see them land in Pinecone. If I don't want that metadata, I have processors to tweak the record before it reaches the destination.

alarbada replied:
thanks for the input @lovromazgon !

alarbada commented on writer_test.go:
I added tests for batching here #22. The WriteDelete integration test should also test for a correct implementation of Writer. 

Generally I try to keep test code minimum while also covering as much as possible. If I can test for a correct implementation of Writer in the destination integration test, I assume that I don't need much testing for the writer, specially since it is a small wrapper around the pinecone.IndexConnection.

I might even remove later on the Writer struct, as it really doesn't do much appart from wrapping some pinecone.Client calls.

alarbada commented:
@hariso last commit kind of replaces the retry logic I added in a later pr. I added it so that the integration tests pass and we finally have a mergeable branch.

hariso commented on tools.go:
```diff
@@ -14,8 +14,9 @@
 //go:build tools
 
-package main
+package pinecone
```
It should be possible to use `main` with the latest version of SDK.

hariso commented on destination_integration_test.go:
```diff
@@ -15,89 +31,224 @@ import (
+}
+
+func teardown(ctx context.Context, is *is.I, resource connectorResource) {
+	err := resource.Teardown(ctx)
 	is.NoErr(err)
```
This can be a one-liner, `is.NoErr(resource.Teardown(ctx))`. Also, the resource is always a destination. With that, there are less reason to introduce a new interface and a new method for just the above.

alarbada replied:
I remember I got the same comment before, sorry! I'll change it now.

hariso commented on utils.go:
```diff
@@ -0,0 +1,215 @@
+	Values  []float32 `json:"values"`
+}
+
+type recordPayload struct {
```
A Pinecone vector would probably be more accurate, no?

hariso commented on utils.go:
The whole file feels a bit overloaded: we have functions to create a new Pinecone index, parse vector data, write batches, build batches... The name indicates it too, so I'd like to suggest to split it up (some of this probably can go to destination.go as well).

hariso reviewed (commented):
Great work @alarbada!

hariso commented on destination_integration_test.go:
```diff
@@ -15,89 +31,234 @@ import (
 }
 
 func TestMain(t *testing.M) {
+	// on development we want to be able to load the pinecone dynamic private
+	// variables into env vars. On CI we load them without an env file, so we
+	// ignore the possible error that godotenv.Load can give.
 	_ = godotenv.Load()
```
Logging the error would be good.

hariso commented on batches_test.go:
```diff
@@ -0,0 +1,114 @@
+
+		is.Equal(len(batches), 5)
+
+		assertUpsertBatch(is, batches[0], "key1")
```
We'd boost our confidence in these tests a lot if we also tested the payloads (vectors, metadata) as well.

alarbada commented on go.mod:
```diff
@@ -1,6 +1,8 @@
 module github.com/conduitio-labs/conduit-connector-pinecone
 
-go 1.22.2
+go 1.21.1
+
+toolchain go1.22.2
```
Nope, first time I see this syntax. I'll change it to 1.22.2

hariso commented on destination_integration_test.go:
```diff
@@ -15,89 +31,237 @@ import (
+		assertDeletedRecordIndex(ctx, t, is, index, id)
+	}
+
+	deleteAllRecords(is, index)
```
Ideally, this is in a defer call, or t.Cleanup.

hariso commented on destination_integration_test.go:
```diff
@@ -15,89 +31,237 @@ import (
-	assertWrittenRecord(is, ctx, index, id, rec)
-}
+	index := createIndex(is, destCfg)
+	defer deleteAllRecords(is, index)
```
Ideally, this is in a defer call, or t.Cleanup.

hariso reviewed (approved):
Thanks @alarbada !
